cryptofeed/exchange/ftx.py

```diff
@@ -81,6 +102,17 @@ async def subscribe(self, conn: AsyncConnection):
             if chan == OPEN_INTEREST:
                 asyncio.create_task(self._open_interest(symbols))  # TODO: use HTTPAsyncConn
                 continue
+            if chan == feed_to_exchange(self.id, UNDERLYING_INDEX):
+                # Construct a list of index symbols
+                index_symbols = []
+                for s in symbols:
+                    if self._is_index(s):
+                        index_symbols.append(s)
+                
+                if index_symbols:
+                    # Create background task to fetch index prices
+                    asyncio.create_task(self._index_price(symbols))
```

Comment:
this works on the happy path but need to check what happens on error...

cryptofeed usually tries to reconnect if the task fails, but this is a newly spawned task so i doubt it has the error handler

at the very least you may need to handle the error in here and raise if something goes wrong

Reply:
yeah doesn't look very happy when the orchestrator sends a msg to drop the FTX index

```
[root] INFO: Received DROP:[('FTX', 'INDEX_BTC')] Signal from NATS...
[asyncio] ERROR: Task exception was never retrieved
future: <Task finished coro=<FTX._subscribe_index_prices() done, defined at /Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/exchange/ftx.py:365> exception=KeyError(('FTX', 'INDEX_BTC')) created at /Users/tmurfitt/opt/miniconda3/envs/elw37/lib/python3.7/asyncio/tasks.py:351>
source_traceback: Object created at (most recent call last):
  File "/Users/tmurfitt/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/211.7142.13/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 2173, in <module>
    main()
  File "/Users/tmurfitt/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/211.7142.13/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 2164, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/Users/tmurfitt/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/211.7142.13/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 1476, in run
    return self._exec(is_module, entry_point_fn, module_name, file, globals, locals)
  File "/Users/tmurfitt/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/211.7142.13/PyCharm.app/Contents/plugins/python/helpers/pydev/pydevd.py", line 1483, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/Users/tmurfitt/Library/Application Support/JetBrains/Toolbox/apps/PyCharm-P/ch-0/211.7142.13/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/Users/tmurfitt/go/src/github.com/elwoodam/market/services/fh_cryptofeed/main.py", line 760, in <module>
    main()
  File "/Users/tmurfitt/go/src/github.com/elwoodam/market/services/fh_cryptofeed/main.py", line 750, in main
    loop.run_forever()
  File "/Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/connection_handler.py", line 63, in _create_connection
    await self.subscribe(connection)
  File "/Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/exchange/ftx.py", line 111, in subscribe
    asyncio.create_task(self._subscribe_index_prices(symbols))
  File "/Users/tmurfitt/opt/miniconda3/envs/elw37/lib/python3.7/asyncio/tasks.py", line 351, in create_task
    return loop.create_task(coro)
Traceback (most recent call last):
  File "/Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/exchange/ftx.py", line 392, in _subscribe_index_prices
    price=Decimal(future['index'])
  File "/Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/feed.py", line 340, in callback
    await cb(**kwargs)
  File "/Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/callback.py", line 100, in __call__
    await super().__call__(feed, symbol, price, timestamp, receipt_timestamp)
  File "/Users/tmurfitt/go/src/github.com/elwoodam/cryptofeed/cryptofeed/callback.py", line 21, in __call__
    await self.callback(*args, **kwargs)
  File "/Users/tmurfitt/go/src/github.com/elwoodam/market/services/fh_cryptofeed/main.py", line 473, in underlying_index
    exchange=SYMBOL_MAP[feed, symbol][0],
KeyError: ('FTX', 'INDEX_BTC')
```

Reply:
^ i think what happened here is the `INDEX_BTC` task wasn't shut down during the normal graceful process, so when that task got the next `INDEX_BTC` price it couldnt find it in our `fh_cryptofeed` map and raised an error.

Reply:
this is the part we want it to use
https://github.com/tristan-murfitt-elw/cryptofeed/blob/0f8eb937bcafb42df683590a94f9cae55dbb587d/cryptofeed/feed.py#L375